File: app/src/main/java/com/droidrank/tictactoe/MainPresenter.java
package com.droidrank.tictactoe;

import com.droidrank.tictactoe.domain.Board;
import com.droidrank.tictactoe.domain.BoardController;

public class MainPresenter implements MainContract.Presenter {

    private MainContract.View view;
    private BoardController boardController;
    private String player1Symbol;
    private String player2Symbol;

    public void setView(MainContract.View view) {
        this.view = view;
    }

    @Override public void initialize() {
        player1Symbol = view.getPlayer1Symbol();
        player2Symbol = view.getPlayer2Symbol();
        boardController = new BoardController(new Board(), player1Symbol, player2Symbol);
        renderBoard();
    }

    @Override public void onResume() {
        // Empty
    }

    @Override public void onPause() {
        // Empty
    }

    @Override public void onDestroy() {
        this.view = null;
    }

    @Override public void onBlockClick(int x, int y) {
        boardController.selectBlock(x, y);
        renderBoard();
    }

    @Override public void onRestartClick() {
        if (view != null) view.showRestartDialog();
    }

    @Override public void restart() {
        boardController.reset();
        renderBoard();
    }

    private void renderBoard() {
        if (view != null) view.render(boardController.getBoard());
    }
}

File: app/src/main/java/com/droidrank/tictactoe/MainPresenter.kt
package com.droidrank.tictactoe

import com.droidrank.tictactoe.domain.Board
import com.droidrank.tictactoe.domain.BoardController
import kotlin.properties.Delegates

class MainPresenter : MainContract.Presenter {

    override var view: MainContract.View? = null
    private lateinit var boardController: BoardController
    private var player1Symbol by Delegates.notNull<String>()
    private var player2Symbol by Delegates.notNull<String>()

    override fun initialize() {
        view?.let {
            player1Symbol = it.player1Symbol
            player2Symbol = it.player2Symbol
            boardController = BoardController(Board(), player1Symbol, player2Symbol)
            it.render(boardController.board)
        }
    }

    override fun onBlockClick(x: Int, y: Int) {
        boardController.selectBlock(x, y)
        view?.render(boardController.board)
    }

    override fun onRestartClick() {
        when {
            boardController.board.isPlayer1Victory -> restart()
            boardController.board.isPlayer2Victory -> restart()
            boardController.board.isDraw -> restart()
            else -> view?.showRestartDialog()
        }
    }

    override fun restart() {
        boardController.reset()
        view?.render(boardController.board)
    }

    override fun onDestroy() {
        this.view = null
    }

}
